db_init.py

```diff
@@ -28,12 +32,48 @@
     UserCourses,
     UserDepartments,
     UserMajors,
-    UserSavedOpportunities,
-    Codes,
+    Codes
 )
 
+
+def fetch_json_data(json_url):
+    response = requests.get(json_url)
+    
+    if response.status_code != 200:
+        raise ValueError(f"Error: Received status code {response.status_code}")
+    try:
+        return response.json()
+    except requests.exceptions.JSONDecodeError:
+        raise ValueError("Error: Received invalid JSON response")
+
+
+def insert_courses_from_json(session, courses_data):
+    # Fetch existing courses to avoid multiple queries
+    existing_courses = {course.code: course for course in session.query(Courses).all()}
+    new_courses = []
+    
+    for course, course_info in courses_data.items():
+        course_name = course_info.get("name")
+        course_code = course_info.get('subj') + course_info.get('crse')
+
+        if len(course_code) != 8:
+            continue
+        if course_code in existing_courses:
+            # Update name if changed
+            existing_course = existing_courses[course_code]
+            if existing_course.name != course_name:
+                existing_course.name = course_name
+        else:
+            new_courses.append(Courses(code=course_code, name=course_name))
+
+    if new_courses:
+        session.add_all(new_courses)
+    session.commit()
```

Comment:
for session.commit, move that into the if statement since we only need it in the if statement